### ISSUE TITLE
Compiler crash when declaring a block receiving an Int

### ISSUE THREAD

**Issue by naps62** [manastech/crystal#335]
I found this problem on an Arch Linux machine, running Crystal 0.5.8.
The function names indicate which methods fail and which ones work

```crystal
class X
  def fail_1(&@b : Int -> _)
    yield 1
  end

  def fail_2(&b : Int -> _)
    @b = b
    yield 1
  end

  def work_1(&@b : Int32 -> _)
    yield 1
  end

  def work_2(&b : Int -> _)
    yield 1
  end
end

X.new.fail_1 { puts "this fails" }
X.new.fail_2 { puts "this fails as well" }
X.new.work_1 { puts "this one works" }
X.new.work_2 { puts "this works as well" }
```

In summary, not using instance variables, or using Int32 instead of Int makes this code work
For the failing ones, what I got was this:

```shell
; ModuleID = 'X'

%String = type { i32, i32, i32, i8 }
%X = type { i32, %"(Nil | (Int -> Nil) | (Int32 -> Nil))" }
%"(Nil | (Int -> Nil) | (Int32 -> Nil))" = type { i32, [2 x i64] }
%"->" = type { i8*, i8* }

@symbol_table = external global [1 x %String*]

define %X* @"*X::new:X"() {
alloca:
  %x = alloca %X*
  br label %entry

entry:                                            ; preds = %alloca
  %0 = call i8* @__crystal_malloc(i32 ptrtoint (%X* getelementptr (%X* null, i32 1) to i32))
  %1 = bitcast i8* %0 to %X*
  %2 = bitcast %X* %1 to i8*
  call void @llvm.memset.p0i8.i32(i8* %2, i8 0, i32 ptrtoint (%X* getelementptr (%X* null, i32 1) to i32), i32 4, i1 false)
  %3 = getelementptr inbounds %X* %1, i32 0, i32 0
  store i32 33, i32* %3
  store %X* %1, %X** %x
  %4 = load %X** %x
  %5 = call i1 @"*GC::add_finalizer<X>:Nil"(%X* %4)
  %6 = load %X** %x
  ret %X* %6
}

declare i8* @__crystal_malloc(i32)

; Function Attrs: nounwind
declare void @llvm.memset.p0i8.i32(i8* nocapture, i8, i32, i32, i1) #0

declare i1 @"*GC::add_finalizer<X>:Nil"(%X*)

define i1 @"*X#fail_1<X, &(Int -> Nil)>:Nil"(%X* %self, %"->" %b) {
alloca:

entry:                                            ; No predecessors!
  %0 = getelementptr inbounds %X* %self, i32 0, i32 1
  %1 = getelementptr inbounds %"(Nil | (Int -> Nil) | (Int32 -> Nil))"* %0, i32 0, i32 0
  store i32 157, i32* %1
  %2 = getelementptr inbounds %"(Nil | (Int -> Nil) | (Int32 -> Nil))"* %0, i32 0, i32 1
  %3 = bitcast [2 x i64]* %2 to %"->"*
  store %"->" %b, %"->"* %3
}

attributes #0 = { nounwind }
```

**Comment by asterite** [comment]
Thanks for reporting!

The main issue is that right now you can't create a function pointer where `Int` is an input. The same happens with Int as a generic type argument: it crashes the compiler. But only because we didn't implement it yet.

All of the comments of [this issue](https://github.com/manastech/crystal/issues/318) apply to this issue as well, so I'll leave both issues open until we decide what to do with `Int`. @waj suggested that we could rename `Int` to `Integer` or something else so users would normally avoid using it (maybe?), but still allow to have `Int` as a generic type argument (although it will be just a bit less efficient).

**Comment by asterite** [comment]
By the way, in `work_1` and `work_2` no function pointer (or Proc) is created because you never capture the block (b), that is, you never refer it by name. In this case Crystal will inline the code and not create a Proc (leads to much more efficient code). That's why it doesn't crash. But in the other cases you are assigning `b` to `@b`, and that creates a Proc and then it crashes.

**Comment by asterite** [comment]
Fixed by 4b1e0742bdae8dabea12ce977ed324339705cd16
